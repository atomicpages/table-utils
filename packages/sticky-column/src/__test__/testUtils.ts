function generate(rows: number, sticky: number[]) {
    const fragment = document.createDocumentFragment();

    for (let i = 0; i < rows; i++) {
        const tr = document.createElement('tr');

        tr.innerHTML = `<td class="${sticky.includes(0) ? 'sticky' : ''}">John Doe</td>
            <td class="${sticky.includes(1) ? 'sticky' : ''}">42</td>
            <td class="${sticky.includes(2) ? 'sticky' : ''}">Male</td>
            <td class="${sticky.includes(3) ? 'sticky' : ''}>000-000-0000</td>
            <td>johndoe@foo.com</td>
            <td>123 Sleepytime Ln.</td>
            <td>San Francisco</td>
            <td>California</td>
            <td>12345</td>`;

        fragment.appendChild(tr);
    }

    return fragment;
}

export function exampleDOM({ rows = 10, sticky = [] }: { rows?: number; sticky?: number[] } = {}) {
    const root = document.createElement('div');
    const table = document.createElement('table');

    table.innerHTML = `<thead>
    <tr>
        <th class="${sticky.includes(0) ? 'sticky' : ''}">Name</th>
        <th class="${sticky.includes(1) ? 'sticky' : ''}">Age</th>
        <th class="${sticky.includes(2) ? 'sticky' : ''}">Gender</th>
        <th class="${sticky.includes(3) ? 'sticky' : ''}">Phone Number</th>
        <th>Email Address</th>
        <th>Address</th>
        <th>City</th>
        <th>State</th>
        <th>Zip</th>
    </tr>
</thead>
<tbody></tbody>`;

    table.appendChild(generate(rows, sticky));
    root.appendChild(table);

    return root;
}

export function mockHTMLProperty(property: keyof HTMLElement, ret: (...args: any[]) => unknown) {
    const mock = jest.fn(ret);

    Object.defineProperty(window.HTMLElement.prototype, property, {
        get: mock,
    });

    const reset = () => {
        mock.mockRestore();
    };

    return { mock, reset };
}
