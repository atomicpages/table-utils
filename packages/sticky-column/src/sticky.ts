import { Options } from '.';

import {
    getStickyCellWidths,
    createGlobalStyles,
    addClassNamesToStickyCells,
    addInlineStylesToStickyCells,
} from './utils';

const normalizeOpts = (options: Options) => {
    if (options.stickyClassName.charAt(0) === '.') {
        options.stickyClassName = options.stickyClassName.slice(1);
    }

    return options;
};

export const sticky = (root: HTMLElement, options?: Options) => {
    const opts = normalizeOpts({
        tr: 'tr',
        thead: 'thead',
        th: 'th',
        stickyClassName: 'sticky',
        ...options,
    });

    let sheet: HTMLStyleElement;
    const stickyCellWidths = getStickyCellWidths(root, opts);

    const cleanup = () => {
        if (opts.globalStylePrefix && sheet) {
            sheet.remove();
        }
    };

    if (opts.globalStylePrefix) {
        sheet = createGlobalStyles(stickyCellWidths, opts);
        addClassNamesToStickyCells(root, opts);
    } else {
        addInlineStylesToStickyCells(root, opts, stickyCellWidths);
    }

    return { cleanup };
};
