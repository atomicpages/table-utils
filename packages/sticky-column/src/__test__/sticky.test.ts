import { sticky } from '../sticky';
import { exampleDOM } from './testUtils';

describe('sticky tests', () => {
    it('should run without errors', () => {
        const root = exampleDOM({ sticky: [0, 1], rows: 2 });

        expect(() => {
            sticky(root);
        }).not.toThrow();
    });

    it('should tolerate . in the stickyClassName', () => {
        const root = exampleDOM({ sticky: [0, 1], rows: 2 });
        sticky(root, { stickyClassName: '.sticky' });
        expect(root).toMatchSnapshot();
    });

    it('should remove internal styles when cleanup is called', () => {
        const root = exampleDOM({ sticky: [0, 1], rows: 2 });
        const { cleanup } = sticky(root, { globalStylePrefix: 'scl' });

        expect(document.head.querySelector('style')).not.toBeNull();
        cleanup();
        expect(document.head.querySelector('style')).toBeNull();
    });

    it('should not throw if cleanup is run', () => {
        const root = exampleDOM({ sticky: [0, 1], rows: 2 });
        const { cleanup } = sticky(root);

        expect(() => {
            cleanup();
        }).not.toThrow();
    });
});
