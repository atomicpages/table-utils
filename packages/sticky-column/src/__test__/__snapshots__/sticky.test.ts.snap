// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sticky tests should tolerate . in the stickyClassName 1`] = `
<div>
  <table>
    <thead>
      
    
      <tr>
        
        
        <th
          class="sticky"
          style="left: 0px;"
        >
          Name
        </th>
        
        
        <th
          class="sticky"
          style="left: 0px;"
        >
          Age
        </th>
        
        
        <th
          class=""
        >
          Gender
        </th>
        
        
        <th
          class=""
        >
          Phone Number
        </th>
        
        
        <th>
          Email Address
        </th>
        
        
        <th>
          Address
        </th>
        
        
        <th>
          City
        </th>
        
        
        <th>
          State
        </th>
        
        
        <th>
          Zip
        </th>
        
    
      </tr>
      

    </thead>
    

    <tbody />
    <tr>
      <td
        class="sticky"
        style="left: 0px;"
      >
        John Doe
      </td>
      
            
      <td
        class="sticky"
        style="left: 0px;"
      >
        42
      </td>
      
            
      <td
        class=""
      >
        Male
      </td>
      
            
    </tr>
    <tr>
      <td
        class="sticky"
        style="left: 0px;"
      >
        John Doe
      </td>
      
            
      <td
        class="sticky"
        style="left: 0px;"
      >
        42
      </td>
      
            
      <td
        class=""
      >
        Male
      </td>
      
            
    </tr>
  </table>
</div>
`;
